python/shotgun_model/shotgun_hierarchy_model.py

```diff
@@ -0,0 +1,855 @@
+        """
+        Asynchronous callback - the data retriever failed to do some work
+
+        :param uid: The unique id of the work that failed
+        :param msg: The error message returned for the failure
+        """
+        uid = sanitize_qt(uid)  # qstring on pyqt, str on pyside
+        msg = sanitize_qt(msg)
+
+        if uid not in self._running_query_lookup:
+            # not our job. ignore
+            logger.debug("Retrieved error from data worker: %s" % (msg,))
+            return
+
+        url = self._running_query_lookup[uid]
+
+        # query is done. clear it out
+        del self._running_query_lookup[uid]
+
+        full_msg = (
+            "Error retrieving data from Shotgun.\n"
+            "  URL: %s\n"
+            "  Error: %s\n"
+        ) % (url, msg)
+        self.data_refresh_fail.emit(full_msg)
+        logger.warning(full_msg)
+
+    def _on_data_retriever_work_completed(self, uid, request_type, data):
+        """
+        Signaled whenever the data retriever completes some work.
+        This method will dispatch the work to different methods
+        depending on what async task has completed.
+
+        :param uid:             The unique id of the work that completed
+        :param request_type:    Type of work completed
+        :param data:            Result of the work
+        """
+        uid = sanitize_qt(uid)  # qstring on pyqt, str on pyside
+        data = sanitize_qt(data)
+
+        logger.debug("Received worker payload of type: %s" % (request_type,))
+
+        if uid not in self._running_query_lookup:
+            # not our job. ignore.
+            return
+
+        # query is done. clear it out
+        del self._running_query_lookup[uid]
+
+        nav_data = data["nav"]
+        self.__on_nav_data_arrived(nav_data)
+
+    def _populate_default_thumbnail(self, item):
+        """
+        Sets the icon for the supplied item based on its "kind" as returned
+        by the ``nav_expand()`` api call.
+
+        :param item: The :class:`~PySide.QtGui.QStandardItem` item to set the
+            icon for.
+        """
+
+        data = get_sg_data(item)
+
+        item_ref = data.get("ref", {})
+        item_kind = item_ref.get("kind")
+
+        if item_kind == "entity":
+            entity_type = item_ref.get("value", {}).get("type")
+            icon = self._shotgun_globals.get_entity_type_icon(entity_type)
+        elif item_kind == "entity_type":
+            entity_type = item_ref["value"]
+            icon = self._shotgun_globals.get_entity_type_icon(entity_type)
+        elif item_kind == "list":
+            icon = self._folder_icon
+        elif item_kind == "no_entity":
+            # this is typically items like "Shots with no Sequence"
+            icon = self._folder_icon
+        else:
+            icon = self._none_icon
+
+        if icon:
+            item.setIcon(icon)
+
+    def _query_hierarchy(self, path, seed_entity_field, entity_fields):
```

Comment:
private? or add example to doc string of when it would be useful to subclass it.